src/Configuration/Programmer/Programmer_cmd.cpp

```diff
@@ -19,57 +19,145 @@ You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <iostream>
+
 #include "CFGCommon/CFGCommon.h"
 #include "Programmer.h"
 
+void progressCallback(std::string msg) { std::cout << msg; }
+
+void messageCallback(std::string msg) { std::cout << "MS-CALLBACK :" << msg; }
+
 using namespace FOEDAG;
 int main(int argc, const char** argv) {
   CFG_POST_MSG("This is Programmer cmd");
   CFGCommon_ARG cmdarg;
 
   // programmer_entry(&cmdarg);
-
   // simple programmer API manual testing
+
+  auto formatMemorySize = [](int sizeInBits) -> std::string {
+    const int bitsInKilobit = 1024;
+    const int bitsInMegabit = 1024 * 1024;
+
+    if (sizeInBits >= bitsInMegabit) {
+      int megabits = sizeInBits / bitsInMegabit;
+      return std::to_string(megabits) + "M";
+    } else if (sizeInBits >= bitsInKilobit) {
+      int kilobits = sizeInBits / bitsInKilobit;
+      return std::to_string(kilobits) + "K";
+    } else {
+      return std::to_string(sizeInBits);
+    }
+  };
+
+  std::string openOCDPath = "/home/guanyung/dev/openocd/src/openocd";
+  openOCDPath = "/home/asic01/development/openocd_rs/src/openocd";
+  int ret = InitLibrary(openOCDPath);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  }
   std::vector<Device> devices;
-  std::string outputMsg;
-  if (ListDevices(devices, outputMsg)) {
+
+  std::vector<FOEDAG::Cable> cables;
+  ret = GetAvailableCables(cables);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  } else if (cables.size() == 0) {
+    CFG_POST_MSG("No programming cable found. Make sure it is connected.");
+    return 0;
+  }
+  CFG_POST_MSG("return %d", ret);
+  CFG_POST_MSG("-- GetAvailableCables(cables) --")
+  for (auto& cable : cables) {
+    CFG_POST_MSG("--------------")
+    CFG_POST_MSG("product ID: 0x%04x", cable.productId);
+    CFG_POST_MSG("vendor  ID: 0x%04x", cable.vendorId);
+    CFG_POST_MSG("serial    : %s", cable.serialNumber.c_str());
+    CFG_POST_MSG("desc      : %s", cable.description.c_str());
+    CFG_POST_MSG("bus       : %d", cable.busAddr);
+    CFG_POST_MSG("port      : %d", cable.portAddr);
+    CFG_POST_MSG("device    : %d", cable.deviceAddr);
+    CFG_POST_MSG("speed     : %d", cable.speed);
+    CFG_POST_MSG("transport : %d", cable.transport);
+    CFG_POST_MSG("channel   : %d", cable.channel);
+  }
+
+  CFG_POST_MSG("-- ListDevices(cables, outputMsg) --")
+  if (cables.size() > 0) {
```

Comment:
Don't we already return at line 68 if cable is not set/found?

Reply:
Yes. I will remove that